### PR TITLE
Optionally parallel downloading csi-provisioner

### PR DESCRIPTION
**Pull request by 0sewa0**
# Description

Introduces a way to run multiple reconciles in parallel in the csi-provisioner, hidden behind an ENVVAR (`MAX_PARALLEL_RECONCILES`).

This is trying to speed up the startup incase multiple dynakubes with different tenants are deployed at the same time all needing the csi driver to mount volumes.

### Edge cases we check for:
- Out of range values for `MAX_PARALLEL_RECONCILES` (should be between 1 - 40)
- We requeue a dynakube reconcile request if:
  - we reached the max defined in the envvar for parallel reconciles running at the same time
  - we are already reconciling the requested dynakube in one of the parallel reconciles

### Gotchas:
- The default memoryLimit and cpuLimit were meant for 1 download at a time and not `x`, so those probably need to be raised at high number of parallel downloads
  - memoryLimit can cause `CrashLoopBackOff` because of `OOMKilled`
  - cpuLimit can cause the downloads to be slow, especially at the unzip part. (the pods wont get killed but kubernetes wont give the necessary cpu to do the job in time)
- Error handling suffers a bit, due to the fact that the framework we use for reconciliation is not meant to be used this way, so it can happen that something is failing in the provisioner yet we can only find it out from the logs of the container. 

## How can this be tested?
1. Have a cluster (that didn't have csi-drivers on them before to make sure that we always have to download the binaries)
2. Modify the csi daemonset yaml to have the `MAX_PARALLEL_RECONCILES` set to something more then "1"
3. Deploy as many dynakubes with different tenants as you can do. (`applicationMonitoring` requires less setup, don't forget to use the `useCSIDriver: true`)
4. Deploy sample apps at the same time for all the dynakubes. (be aware of the `namespaceSelector`)

Expected result: 
Parallel downloads for each tenant (grep for `parallel` in the logs)

## Checklist
- [x] Unit tests have been updated/added 
- [x] PR is labeled accordingly

### CODE REVIEW

File: src/controllers/csi/provisioner/controller.go [hunk 1]
```diff
@@ -49,23 +50,25 @@ type OneAgentProvisioner struct {
 	apiReader    client.Reader
 	opts         dtcsi.CSIOptions
 	dtcBuildFunc dynakube.DynatraceClientFunc
-	fs           afero.Fs
 	recorder     record.EventRecorder
 	db           metadata.Access
 	path         metadata.PathResolver
+
+	currentParallel int64
```
Comment (mjgrzybek) [inline src/controllers/csi/provisioner/controller.go:57]:
```suggestion
	currentParallelDownloads int64
```

File: src/controllers/csi/provisioner/controller.go [hunk 2]
```diff
@@ -49,23 +50,25 @@ type OneAgentProvisioner struct {
...
+	maxParallel     int64
```
Comment (mjgrzybek) [inline src/controllers/csi/provisioner/controller.go:58]:
```suggestion
	maxParallelDownloads     int64
```

File: src/controllers/csi/provisioner/controller.go [hunk 3]
```diff
@@ -76,6 +79,27 @@ func (provisioner *OneAgentProvisioner) SetupWithManager(mgr ctrl.Manager) error
 }
 
 func (provisioner *OneAgentProvisioner) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	fs := afero.NewOsFs()
+	if provisioner.maxParallel > 0 {
+		if provisioner.currentParallel == provisioner.maxParallel {
+			log.Info("max parallel downloads reached, requeued")
+			return reconcile.Result{RequeueAfter: shortRequeueDuration}, nil
+		}
```
Comment (mjgrzybek) [inline src/controllers/csi/provisioner/controller.go:87]:
```suggestion

func (provisioner *OneAgentProvisioner) eligibleToParallelDownloads() bool {
	return provisioner.maxParallel > 0
}


func (provisioner *OneAgentProvisioner) parallelDownloadsLimitReached() bool {
	return provisioner. provisioner.currentParallel >= provisioner.maxParallel
}

func (provisioner *OneAgentProvisioner) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
	fs := afero.NewOsFs()
	if provisioner.maxParallel > 0 {
		if provisioner.eligibleToParallelDownloads() {
			provisioner.parallelDownloadsLimitReached() {
				log.Info("max parallel downloads reached, requeued")
				return reconcile.Result{RequeueAfter: shortRequeueDuration}, nil
			}
```

File: src/controllers/csi/provisioner/controller.go [hunk 4]
```diff
@@ -76,6 +79,27 @@ func (provisioner *OneAgentProvisioner) SetupWithManager(mgr ctrl.Manager) error
 }
 
 func (provisioner *OneAgentProvisioner) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	fs := afero.NewOsFs()
+	if provisioner.maxParallel > 0 {
+		if provisioner.currentParallel == provisioner.maxParallel {
```
Comment (mjgrzybek) [inline src/controllers/csi/provisioner/controller.go:84]:
```suggestion
		if provisioner.currentParallel >= provisioner.maxParallel {
```
a bit more defensive condition :slightly_smiling_face: .

File: src/controllers/csi/provisioner/controller.go [hunk 5]
```diff
@@ -76,6 +79,27 @@ func (provisioner *OneAgentProvisioner) SetupWithManager(mgr ctrl.Manager) error
...
+		}
+		atomic.AddInt64(&provisioner.currentParallel, 1)
+		log.Info("staring parallel download")
+		go func() {
+			_, err := provisioner.reconcile(ctx, request, fs)
+			if err != nil {
+				log.Error(err, "Problem while provisioning oneagents in parallel")
+			}
+			atomic.AddInt64(&provisioner.currentParallel, -1)
+		}()
```
Comment (mjgrzybek) [inline src/controllers/csi/provisioner/controller.go:96]: if `provisioner.reconcile()` panicked, `currentParallel` wouldn't decrease :boom:.
```suggestion
		go func() {
			defer atomic.AddInt64(&provisioner.currentParallel, -1)
			_, err := provisioner.reconcile(ctx, request, fs)
			if err != nil {
				log.Error(err, "Problem while provisioning oneagents in parallel")
			}
		}()
```

File: src/controllers/csi/provisioner/controller.go [hunk 6]
```diff
@@ -49,23 +50,25 @@ type OneAgentProvisioner struct {
...
 	db           metadata.Access
 	path         metadata.PathResolver
+
+	currentParallel int64
+	maxParallel     int64
 }
 
 // NewOneAgentProvisioner returns a new OneAgentProvisioner
-func NewOneAgentProvisioner(mgr manager.Manager, opts dtcsi.CSIOptions, db metadata.Access) *OneAgentProvisioner {
+func NewOneAgentProvisioner(mgr manager.Manager, opts dtcsi.CSIOptions, db metadata.Access, maxParallel int64) *OneAgentProvisioner {
```
Comment (mjgrzybek) [inline src/controllers/csi/provisioner/controller.go:62]:
```suggestion
func NewOneAgentProvisioner(mgr manager.Manager, opts dtcsi.CSIOptions, db metadata.Access, maxParallelDownloads int64) *OneAgentProvisioner {
```

File: src/cmd/operator/csi_driver.go [hunk 7]
```diff
@@ -95,7 +96,8 @@ func setupCSIDriver(ns string, cfg *rest.Config) (manager.Manager, func(), error
 		return nil, cleanUp, err
 	}
 
-	if err := csiprovisioner.NewOneAgentProvisioner(mgr, csiOpts, access).SetupWithManager(mgr); err != nil {
+	maxParallel, _ := strconv.ParseInt(os.Getenv("MAX_PARALLEL"), 10, 64)
```
Comment (mjgrzybek) [inline src/cmd/operator/csi_driver.go:99]:
```suggestion
	maxParallel, _ := strconv.ParseInt(os.Getenv("MAX_PARALLEL_DOWNLOADS"), 10, 64)
```

File: src/cmd/operator/csi_driver.go [hunk 8]
```diff
@@ -95,7 +96,8 @@ func setupCSIDriver(ns string, cfg *rest.Config) (manager.Manager, func(), error
 		return nil, cleanUp, err
 	}
 
-	if err := csiprovisioner.NewOneAgentProvisioner(mgr, csiOpts, access).SetupWithManager(mgr); err != nil {
+	maxParallelDownloads, _ := strconv.ParseInt(os.Getenv("MAX_PARALLEL_DOWNLOADS"), 10, 64)
```
Comment (mjgrzybek) [inline src/cmd/operator/csi_driver.go:99]: Error handling is necessary here (testcase: `MAX_PARALLEL_DOWNLOADS=`:trollface:)

File: src/controllers/csi/provisioner/controller.go [hunk 9]
```diff
@@ -76,6 +83,34 @@ func (provisioner *OneAgentProvisioner) SetupWithManager(mgr ctrl.Manager) error
...
 
 func (provisioner *OneAgentProvisioner) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	fs := afero.NewOsFs()
+	if provisioner.parallelDownloadsEnabled() {
+		if provisioner.parallelDownloadsLimitReached() {
+			log.Info("max parallel downloads reached, requeued")
+			return reconcile.Result{RequeueAfter: shortRequeueDuration}, nil
+		}
+		atomic.AddInt64(&provisioner.currentParallelDownloads, 1)
+		log.Info("staring parallel download")
```
Comment (mjgrzybek) [inline src/controllers/csi/provisioner/controller.go:93]:
```suggestion
		log.Info("starting parallel download")
```

File: src/cmd/operator/csi_driver.go (hunk 8 above)
Comment (meik99) [inline src/cmd/operator/csi_driver.go:99]: please don't inline strings like these, refactor it into a constant.

Comment (meik99) [reply src/cmd/operator/csi_driver.go:99]: Alongside the error handling I would appreciate sane limits, i.e., min and max values, to handle cases like "-1".
All of this could be handled in a function which reads the env var, parses it into an int, checks the value, enforces the limits and then returns the final value

File: src/cmd/operator/csi_driver.go [hunk 10]
```diff
@@ -95,7 +96,8 @@ func setupCSIDriver(ns string, cfg *rest.Config) (manager.Manager, func(), error
...
+	if err := csiprovisioner.NewOneAgentProvisioner(mgr, csiOpts, access, maxParallelDownloads).SetupWithManager(mgr); err != nil {
```
Comment (meik99) [inline src/cmd/operator/csi_driver.go:100]: Could `maxParallelDownloads` be merged into the struct behind csiOpts?

File: src/controllers/csi/provisioner/controller.go [hunk 11]
```diff
@@ -49,23 +50,29 @@ type OneAgentProvisioner struct {
...
-		dtcBuildFunc: dynakube.BuildDynatraceClient,
-		fs:           afero.NewOsFs(),
-		recorder:     mgr.GetEventRecorderFor("OneAgentProvisioner"),
-		db:           db,
-		path:         metadata.PathResolver{RootDir: opts.RootDir},
+		client:               mgr.GetClient(),
+		apiReader:            mgr.GetAPIReader(),
+		opts:                 opts,
+		dtcBuildFunc:         dynakube.BuildDynatraceClient,
+		recorder:             mgr.GetEventRecorderFor("OneAgentProvisioner"),
```
Comment (meik99) [inline src/controllers/csi/provisioner/controller.go:72]: Please refactor this string also to a constant

File: src/controllers/csi/provisioner/controller.go [hunk 12]
```diff
@@ -76,6 +83,34 @@ func (provisioner *OneAgentProvisioner) SetupWithManager(mgr ctrl.Manager) error
...
+		}
+		atomic.AddInt64(&provisioner.currentParallelDownloads, 1)
+		log.Info("staring parallel download")
+		go func() {
+			defer atomic.AddInt64(&provisioner.currentParallelDownloads, -1)
+			_, err := provisioner.reconcile(ctx, request, fs)
+			if err != nil {
+				log.Error(err, "Problem while provisioning oneagents in parallel")
+			}
+		}()
```
Comment (meik99) [inline src/controllers/csi/provisioner/controller.go:100]: Reafctor into a function like and call as `go functionName()`

File: src/controllers/csi/provisioner/controller.go [hunk 13]
```diff
@@ -76,6 +83,34 @@ func (provisioner *OneAgentProvisioner) SetupWithManager(mgr ctrl.Manager) error
...
+}
+
+func (provisioner *OneAgentProvisioner) parallelDownloadsEnabled() bool {
+	return provisioner.maxParallelDownloads > 0
+}
+func (provisioner *OneAgentProvisioner) parallelDownloadsLimitReached() bool {
+	return provisioner.currentParallelDownloads >= provisioner.maxParallelDownloads
+}
+
+func (provisioner *OneAgentProvisioner) reconcile(ctx context.Context, request reconcile.Request, fs afero.Fs) (reconcile.Result, error) {
```
Comment (meik99) [inline src/controllers/csi/provisioner/controller.go:113]: Why is the filesystem not part of the provisioner anymore?

**Review by meik99 — CHANGES_REQUESTED** [review]

File: src/controllers/csi/provisioner/controller.go (hunk 13 above)
Comment (0sewa0) [reply src/controllers/csi/provisioner/controller.go:113]: `afero` is not thread safe, (I don't know how or for which functions)
so to be safe each goroutine gets its own `fs` object

File: src/controllers/csi/provisioner/controller.go [hunk 14]
```diff
@@ -76,6 +83,34 @@ func (provisioner *OneAgentProvisioner) SetupWithManager(mgr ctrl.Manager) error
...
 func (provisioner *OneAgentProvisioner) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	fs := afero.NewOsFs()
+	if provisioner.parallelDownloadsEnabled() {
+		if provisioner.parallelDownloadsLimitReached() {
+			log.Info("max parallel downloads reached, requeued")
+			return reconcile.Result{RequeueAfter: shortRequeueDuration}, nil
+		}
+		atomic.AddInt64(&provisioner.currentParallelDownloads, 1)
+		log.Info("staring parallel download")
+		go func() {
```
Comment (mjgrzybek) [inline src/controllers/csi/provisioner/controller.go:94]: What if reconciliation failed?

Comment (0sewa0) [reply src/controllers/csi/provisioner/controller.go:94]: we will be sad, but we don't care, the logs will be still there

(this is not meant to be a permanent long-term solution and won't be documented so not everyone will use it)

Comment (mjgrzybek) [reply src/controllers/csi/provisioner/controller.go:94]: It changes the logic of `OneAgentProvisioner`. [Currently](https://github.com/Dynatrace/dynatrace-operator/pull/656/files/d2fa7a67887e6ec71a4db5032baccfee711ccff1#diff-f2b8b611648fec932e0790d9a1ae99ad30a8ca183edce27761bbe1271eb71b89L138) it returns an `error` so that request is re-queued to be processed again.
I'd suggest to preserve this logic by adding an atomic bool which indicates if error occurred in any of goroutines.
If that bool==true, return `reconcile.Result{}, errors.New("parallel download failed")`

Comment (mjgrzybek) [inline src/controllers/csi/provisioner/controller.go:94]: What if download is still in progress and another reconciliation started?
Add `sync.WorkGroup` to wait for downloads to be finished before exiting reconciliation loop.

Comment (0sewa0) [reply src/controllers/csi/provisioner/controller.go:94]: that is not how it works 😅

Comment (0sewa0) [reply src/controllers/csi/provisioner/controller.go:94]: we would need a sync group per dynakube, and not block but requeue them

Comment (mjgrzybek) [reply src/controllers/csi/provisioner/controller.go:94]: If I understand it correctly, `provisioner.currentParallelDownloads` is a state shared between reconciler loops, so it may happen that two downloads for single tenant are in progress :thinking:.

Comment (mjgrzybek) [reply src/controllers/csi/provisioner/controller.go:94]: `Reconcile()` should not exit until downloads are done.

Comment (0sewa0) [reply src/controllers/csi/provisioner/controller.go:94]: we would need to block(which we can't) OR log the error in another reconcile, but because we are running stuff in parallel we might override that error, or need to store multiple of these bools.

this shouldn't be over-engineered

File: src/controllers/csi/provisioner/controller.go [hunk 15]
```diff
@@ -49,33 +53,62 @@ type OneAgentProvisioner struct {
 	apiReader    client.Reader
 	opts         dtcsi.CSIOptions
 	dtcBuildFunc dynakube.DynatraceClientFunc
-	fs           afero.Fs
 	recorder     record.EventRecorder
 	db           metadata.Access
 	path         metadata.PathResolver
+
+	mu                       *sync.Mutex
+	currentParallelDownloads int64
```
Comment (0sewa0) [inline src/controllers/csi/provisioner/controller.go:61]: maybe this is redundant now ? 
could just check the len of the map

File: src/controllers/csi/provisioner/controller.go [hunk 16]
```diff
@@ -89,72 +122,110 @@ func (provisioner *OneAgentProvisioner) Reconcile(ctx context.Context, request r
...
 		return reconcile.Result{RequeueAfter: shortRequeueDuration}, nil
 	}
 
+	if provisioner.parallelDownloadsEnabled() {
+		if provisioner.parallelDownloadsLimitReached() {
+			log.Info("max parallel downloads reached, requeued")
+			return reconcile.Result{RequeueAfter: shortRequeueDuration}, nil
+		}
+		if provisioner.isCurrentlyDownloading(tenant) {
+			log.Info("still downloading", "tenant", tenant)
```
Comment (mjgrzybek) [inline src/controllers/csi/provisioner/controller.go:137]: nit: it should be able to cancel download it it's stalled using `context.Context`, but that should be done in another PR

File: src/controllers/csi/provisioner/controller.go [hunk 17]
```diff
@@ -49,33 +53,62 @@ type OneAgentProvisioner struct {
...
+		dtcBuildFunc:         dynakube.BuildDynatraceClient,
+		recorder:             mgr.GetEventRecorderFor(eventRecorderName),
+		db:                   db,
+		path:                 metadata.PathResolver{RootDir: opts.RootDir},
+		mu:                   &sync.Mutex{},
+		currentlyDownloading: map[string]bool{},
 	}
 }
 
+func (provisioner *OneAgentProvisioner) setCurrentlyDownloading(name string) {
```
Comment (mjgrzybek) [inline src/controllers/csi/provisioner/controller.go:83]:
```suggestion
func (provisioner *OneAgentProvisioner) setCurrentlyDownloading(tenant string) {
```

File: src/controllers/csi/provisioner/controller.go (hunk 16 above)
Comment (0sewa0) [reply src/controllers/csi/provisioner/controller.go:137]: that would require that the download would care for that context but it doesn't (that part of the code was written before my time sooo 😬 )

File: src/controllers/csi/provisioner/controller.go [hunk 18]
```diff
@@ -49,33 +53,62 @@ type OneAgentProvisioner struct {
...
-		path:         metadata.PathResolver{RootDir: opts.RootDir},
+		client:               mgr.GetClient(),
+		apiReader:            mgr.GetAPIReader(),
+		opts:                 opts,
+		dtcBuildFunc:         dynakube.BuildDynatraceClient,
+		recorder:             mgr.GetEventRecorderFor(eventRecorderName),
+		db:                   db,
+		path:                 metadata.PathResolver{RootDir: opts.RootDir},
+		mu:                   &sync.Mutex{},
+		currentlyDownloading: map[string]bool{},
```
Comment (mjgrzybek) [inline src/controllers/csi/provisioner/controller.go:79]: Map's values are not needed here. We need a Set, which can be created this way:
```suggestion
		currentlyDownloading: map[string]interface{},
```

File: src/controllers/csi/provisioner/controller.go (hunk 15 above)
Comment (mjgrzybek) [reply src/controllers/csi/provisioner/controller.go:61]: see: https://github.com/Dynatrace/dynatrace-operator/pull/656#discussion_r831995831

File: src/controllers/csi/provisioner/controller.go (hunk 18 above)
Comment (0sewa0) [reply src/controllers/csi/provisioner/controller.go:79]: empty interface gives us nothing, just longer lines of code

also the correct way is `currentlyDownloading: map[string]interface{}{}` which is ugly as sin 😬

Comment (0sewa0) [reply src/controllers/csi/provisioner/controller.go:77]: we don't use a loop(I delete stuff now), and first result for `golang set` from Google is using a map of bools, so the internet is with me 😛

Comment (mjgrzybek) [reply src/controllers/csi/provisioner/controller.go:77]: your google search results are profiled up to your needs :stuck_out_tongue:

Comment (mjgrzybek) [reply src/controllers/csi/provisioner/controller.go:77]: uh, sorry, it's `struct{}`, not `interface{}`!
see: https://www.davidkaya.com/sets-in-golang/

Comment (0sewa0) [reply src/controllers/csi/provisioner/controller.go:77]: that is the 4. result, therefore I still prefer bool 😬

File: src/controllers/csi/provisioner/controller.go [hunk 19]
```diff
@@ -57,9 +57,8 @@ type OneAgentProvisioner struct {
 	db           metadata.Access
 	path         metadata.PathResolver
 
-	mu                       *sync.Mutex
-	currentParallelDownloads int64
-	currentlyDownloading     map[string]bool
+	mu                   *sync.Mutex
+	currentlyReconciling *kubeobjects.StringSet
```
Comment (mjgrzybek) [inline src/controllers/csi/provisioner/controller.go:61]: Why is `StringSet` in `kubeobjects`?

Comment (0sewa0) [reply src/controllers/csi/provisioner/controller.go:61]: I'm open for suggestion 😄

Comment (mjgrzybek) [reply src/controllers/csi/provisioner/controller.go:61]: `utils` ofc. :smile:

Comment (mjgrzybek) [reply src/controllers/csi/provisioner/controller.go:61]: `collections` maybe?

Comment (0sewa0) [reply src/controllers/csi/provisioner/controller.go:61]: oh no 😨

Comment (meik99) [reply src/controllers/csi/provisioner/controller.go:61]: So, a map which only purpose is to collect keys to more easily remove them, since golang has no `remove` function for elements in arrays, now has it's own type which, in turn, does nothing but exist?

That seems highly unnecessary and IMO it should just be `map[string]bool`.

Also, please rename `mu` to `mutex`-

Comment (0sewa0) [reply src/controllers/csi/provisioner/controller.go:61]: 🍿

Comment (0sewa0) [reply src/controllers/csi/provisioner/controller.go:61]: have it in `dtcsi` and call it `TenantsSet` ? would that work ?
(I like the way it simplifies the code of the provisioner even if just a bit, and the logic is simple so I would keep it)

File: src/controllers/csi/provisioner/controller.go [hunk 20]
```diff
@@ -49,38 +53,72 @@ type OneAgentProvisioner struct {
 	apiReader    client.Reader
 	opts         dtcsi.CSIOptions
 	dtcBuildFunc dynakube.DynatraceClientFunc
-	fs           afero.Fs
 	recorder     record.EventRecorder
 	db           metadata.Access
 	path         metadata.PathResolver
+
+	mu                   *sync.Mutex
```
Comment (meik99) [inline src/controllers/csi/provisioner/controller.go:60]:
```suggestion
	mutex                   *sync.Mutex
```

**Review by meik99 — CHANGES_REQUESTED** [review]

**Review by mjgrzybek — APPROVED** [review]
